eos-updater-avahi/eos-updater-avahi.c

```diff
@@ -0,0 +1,320 @@
+
+  /* Work out what commit we would advertise. Errors here are not fatal, as we
+   * want to delete the file on failure. */
+  sysroot = ostree_sysroot_new_default ();
+
+  if (!ostree_sysroot_load (sysroot, NULL, &local_error))
+    {
+      g_warning ("Error loading sysroot: %s", local_error->message);
+      g_clear_error (&local_error);
+    }
+  else if (!eos_sysroot_get_advertisable_commit (sysroot,
+                                                 &commit_checksum,
+                                                 &commit_ostree_path,
+                                                 &commit_timestamp,
+                                                 &local_error))
+    {
+      g_warning ("Error getting advertisable commit: %s", local_error->message);
+      g_clear_error (&local_error);
+    }
+
+  if (commit_checksum != NULL)
+    {
+      commit_date_time = g_date_time_new_from_unix_utc (commit_timestamp);
+      formatted_commit_date_time = g_date_time_format (commit_date_time,
+                                                       "%FT%T%:z");
+    }
+
+  /* Work out the update policy. */
+  if (!advertise_updates && commit_checksum != NULL)
+    {
+      g_message ("Advertising updates is disabled. Deployed commit ‘%s’ "
+                 "(%s, %s) will not be advertised.",
+                 commit_checksum, formatted_commit_date_time,
+                 commit_ostree_path);
+      delete = TRUE;
+    }
+  else if (advertise_updates && commit_checksum == NULL)
+    {
+      g_message ("Advertising updates is enabled, but no appropriate deployed "
+                 "commits were found. Not advertising updates.");
+      delete = TRUE;
+    }
+  else if (!advertise_updates && commit_checksum == NULL)
+    {
+      g_message ("Advertising updates is disabled, and no appropriate deployed "
+                 "commits were found. Not advertising updates.");
+      delete = TRUE;
+    }
+  else
+    {
+      g_message ("Advertising updates is enabled, and deployed commit ‘%s’ "
+                 "(%s, %s) will be advertised.", commit_checksum,
+                 formatted_commit_date_time, commit_ostree_path);
+      delete = FALSE;
+    }
+
+  /* Apply the policy. */
+  if (!delete)
+    {
+      if (!eos_avahi_service_file_generate (avahi_service_directory,
+                                            commit_ostree_path,
+                                            commit_date_time,
+                                            cancellable,
+                                            error))
+        {
+          eos_avahi_service_file_delete (avahi_service_directory, cancellable,
+                                         NULL);
+          return FALSE;
+        }
+
+      return TRUE;
+    }
+  else
+    {
+      return eos_avahi_service_file_delete (avahi_service_directory,
+                                            cancellable, error);
+    }
+}
+
+/* main() exit codes. */
+enum
+{
+  EXIT_SUCCESS = 0,
+  EXIT_FAILURE = 1,
```

Comment:
Yes, but I deliberately put them in here to be explicit about it, since the integration tests rely on the integer values of the exit statuses. I guess we will have problems if `#include <stdlib.h>` sneaks in anywhere, since it `#define`s them rather than `enum`s them. I’ll rename them.